src/main/scala/util/SelectDiplomacy.scala

```diff
@@ -0,0 +1,407 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import scala.io.Source
+import java.nio.file.{Paths, Files}
+
+import Chisel._
+import chisel3.RawModule
+//import chisel3.internal.firrtl.{Circuit, DefModule}
+import chisel3.internal.firrtl._
+import java.io.{File, FileWriter}
+
+import firrtl.annotations.JsonProtocol
+import freechips.rocketchip.config._
+import freechips.rocketchip.system.{DefaultTestSuites, TestGeneration}
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.rocket._
+import freechips.rocketchip.diplomacy.{LazyModuleImp, LazyModule, BaseNode}
+import freechips.rocketchip.diplomaticobjectmodel.HasLogicalTreeNode
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree._
+import freechips.rocketchip.diplomaticobjectmodel.model.{OMComponent, OMInterrupt}
+
+import chisel3.aop.{Aspect, Select}
+
+import firrtl.graph._
+
+import scala.collection.mutable
+
+import scala.collection.JavaConverters._
+
+trait GraphType
+object DOTGRAPH extends GraphType
+object GRAPHML extends GraphType
+
+case class RTreeData (val item: Any)
+
+
+object SelectDiplomacy {
+
+  val digraph = new MutableDiGraph[RTreeData]
+
+  def findNode(name:String): Any = {
+    None
+  }
+
+  /** Collects all Lazy Module Imp as a List.
+    */
+  def collectImpModules(): List[LazyModuleImp] = {
```

Comment:
Many of your collect functions share similar structure, but do not directly reuse code. They all iterate of teh edges, match on the parent.item, and then return something. Can you capture that pattern in a separate function which is called by these collection functions?